### PR TITLE
[FIX] sale_coupon_advanced: order_count

### PR DESCRIPTION
**Pull request by oerp-odoo**
Issue 194

There is already order_count field that shows how many orders have used
related coupon, so it does not make sense to use _get_order_count method
which just returns customer orders, not the one that used coupon.

### CODE REVIEW

**Comment by i-vyshnevska** [comment]
you cannot use order_count as it counts orders for all customers and the method should return the amount of orders for certain customers

**Comment by oerp-odoo** [comment]
Oh, I thought partner did not matter, only used up coupons did. Also there were no tests to check this part, so assumed it was not important. Will redo.

**Comment by oerp-odoo** [comment]
@i-vyshnevska updated PR

File: sale_coupon_advanced/models/sale_coupon_program.py [hunk 1]
```diff
@@ -72,49 +69,60 @@ def _force_sale_order_lines(self, programs, order):
             ):
                 order.add_reward_line_values(program)
 
+    # FIXME: whats this? Original method does not exist on this model.
```
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_coupon_program.py]: pls move method to sale.order model

File: sale_coupon_advanced/views/sale_coupon.program.xml [hunk 2]
```diff
@@ -11,7 +11,7 @@
                 <field name="is_cumulative" />
                 <field name="first_order_only" />
                 <field
-                    name="first_n_customer_orders"
+                    name="next_n_customer_orders"
                     attrs="{'invisible': [('first_order_only', '=', True)]}"
```
Comment (i-vyshnevska) [inline sale_coupon_advanced/views/sale_coupon.program.xml:18]:
```suggestion
```
now this is unrelated features

File: sale_coupon_advanced/models/sale_coupon_program.py [hunk 3]
```diff
@@ -28,18 +28,16 @@ class SaleCouponProgram(models.Model):
...
-
         order_count = self._get_order_count(order)
         if self.first_order_only and order_count:
             return {"error": _("Coupon can be used only for the first sale order!")}
-        max_order_number = self.first_n_customer_orders
+        max_order_number = self.next_n_customer_orders
         if max_order_number and order_count >= max_order_number:
             return {
                 "error": _(
                     "Coupon can be used only for the first {} sale order!"
```
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_coupon_program.py]:
```suggestion
                    "Coupon can be used only for the {} times!"
```

File: sale_coupon_advanced/models/sale_coupon_program.py [hunk 4]
```diff
@@ -53,12 +51,11 @@ def _check_promo_code(self, order, coupon_code):
 
     @api.model
     def _filter_programs_from_common_rules(self, order, next_order=False):
-
         initial_programs = self.browse(self.ids)
         self._force_sale_order_lines(initial_programs, order)
         programs = super()._filter_programs_from_common_rules(order, next_order)
-        programs = programs._filter_first_order_programs(order)
-        programs = programs._filter_n_first_order_programs(order)
+        self._filter_order_programs(order, self._filter_first_order_programs)
```
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_coupon_program.py]: assigning programs here we trying to follow policy defined in the base method especially when we use a model decorator

File: sale_coupon_advanced/models/sale_coupon_program.py (hunk 1 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_coupon_program.py]: Why move? This method does not do anything. And there is already implementation on sale.order anyway.

File: sale_coupon_advanced/models/sale_coupon_program.py (hunk 4 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_coupon_program.py]: You mean, that I forgot to use ` programs` and used `self? Then yeah, will change that.

File: sale_coupon_advanced/models/sale_coupon_program.py (hunk 3 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_coupon_program.py]: If we remove `sale order`, then `the` should be omitted.

Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_coupon_program.py]: Or not, read different part..:)

File: sale_coupon_advanced/views/sale_coupon.program.xml (hunk 2 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/views/sale_coupon.program.xml:18]: Technically those two can work together, but does it make sense from business perspective?

File: sale_coupon_advanced/models/sale_coupon_program.py (hunk 1 above)
Comment (i-vyshnevska) [reply sale_coupon_advanced/models/sale_coupon_program.py]: it is a notice for me, we left some not finished work and it is defined in sale.order in core

File: sale_coupon_advanced/views/sale_coupon.program.xml (hunk 2 above)
Comment (i-vyshnevska) [reply sale_coupon_advanced/views/sale_coupon.program.xml:18]: well yeah they meaningless together, than you need
` attrs="{'invisible': [('first_order_only', '=', False)]}"`
to hide one when the first is raised

**Review by i-vyshnevska — APPROVED** [review]
pre-approving only some nitpicking

**Comment by sebalix** [comment]
Side-note: it's preferable to not use internal tickets references on public repositories :)
EDIT: this include also the branch names

File: sale_coupon_advanced/views/sale_coupon.program.xml (hunk 2 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/views/sale_coupon.program.xml:18]: Well if you mean to put this on `first_order_only`, it would hide it permanently. Something like this should work: 
`attrs="{'invisible': [('first_order_only', '=', True)]}" on next_n_customer_orders`
`attrs="{'invisible': [('next_n_customer_orders', '=', True)]}" on first_order_only`

File: sale_coupon_advanced/models/sale_coupon_program.py (hunk 1 above)
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_coupon_program.py]: Oh, u meant move your TODO comment. OK
